### ISSUE TITLE
Inference speed regression from HistogramAggregator

### ISSUE THREAD

**Issue by brianhenn** [ai2cm/ace#312]
See wandb report [here](https://wandb.ai/ai2cm/fourcastnet/reports/Inline-inference-speed--Vmlldzo3MzA4MzU4). There appears to have been a ~5x slowdown in 1deg inference speed with adding the `HistogramAggregators` (#https://github.com/ai2cm/full-model/pull/673); all of the slowdown was from the aggregator/writer portion which slowed down by ~100x, and the slowdown must come from the call of `HistogramAggregator.record_batch()`.

Would suggest either reverting ai2cm/full-model#673 or making `HistogramAggregator` off by default. Maybe profiling can speed it up, or it can record only 1/1000 of the data or something like that, and we can bring it back.

In general probably a good idea to run some kind of speed test when code changes are made to inference. Could we automate this?

**Comment by oliverwm1** [comment]
Interesting that this is so slow. I had a hunch that the slowness from the histogram data writer was from the actual writing to disk, but it seems like just computing the histogram is quite expensive too, if I understand correctly.

**Comment by brianhenn** [comment]
That's right, the slowdown is shown during time stepping of inline inference, where there's no plotting or data writing involved. Speculatively it seems like the bottleneck is just from calling `np.histogram` on a very long 1D array [here](https://github.com/ai2cm/full-model/blob/2811429373c366144312631a24097024af42dcb9/fme/fme/core/histogram.py#L87C1-L91C10).

**Comment by oliverwm1** [comment]
I wonder if computing it directly from the data on GPU with `torch.histogram` would help. Probably not... I expect the bottleneck is actually from the histogram calculation (versus moving data from GPU to CPU). But could be worth trying. @mcgibbon

**Comment by mcgibbon** [comment]
FWIW the histogram is a very interesting/useful metric to have around, for example check out the upper atmospheric temperature tails in these C24-shield trained runs: https://wandb.ai/ai2cm/fourcastnet/runs/w5i227dl

I'll see if I can find a way to speed it up, it doesn't feel like calculating a histogram should be a bottleneck.

**Comment by mcgibbon** [comment]
~~Would we mind introducing a [new (pretty minimal) dependency](https://pypi.org/project/fast-histogram/) that provides a faster histogram calculation, specific to regularly spaced bins?~~

> The result is a 1D histogram function here that is 7-15x faster than numpy.histogram

Update: unfortunately this dependency doesn't let you specify your own bin start and size, so we can't use it

**Comment by mcgibbon** [comment]
@brianhenn could you re-run your timing using the branch for https://github.com/ai2cm/full-model/pull/700? A speed-up would confirm that the issue is the histogram calculation itself and not the surrounding business logic.

**Comment by brianhenn** [comment]
@mcgibbon I added that run to the [report](https://wandb.ai/ai2cm/fourcastnet/reports/Inline-inference-speed--Vmlldzo3NDAyNjM4/edit). It does help a lot -- with that PR, inference is about 80% as fast as the case of having the HistogramAggregator off entirely as opposed current main that is only ~25% as fast. 

That being said with the PR we're only about 60% as fast as we were on main a couple weeks ago (due to both the histograms and seasonal aggregators adding time). Not sure what the right tradeoff is.

**Comment by mcgibbon** [comment]
@brianhenn I have another PR/branch that also computes the histograms on GPU, if you can check whether that is even faster (or slower) https://github.com/ai2cm/full-model/pull/714.

I think the optimal tradeoff depends on the experiment. For my 4-degree experiments, 10 year inline inference x 4 ensemble members only takes ~20% of the total runtime, and that's before any speed-ups from open PRs. Of course, any easy optimization we can do to make inference faster is a good thing.

**Comment by brianhenn** [comment]
Will do as soon as Perlmutter nodes are un-cordoned this evening...since I've been doing the speed testing there I don't want to switch to another compute environment and confuse the results.

**Comment by brianhenn** [comment]
Added...some further improvement from ai2cm/full-model#714 over ai2cm/full-model#700.

**Comment by mcgibbon** [comment]
@brianhenn I don't seem to have access to [the report](https://wandb.ai/ai2cm/fourcastnet/reports/Inline-inference-speed--Vmlldzo3NDAyNjM4/edit). I think you need to publish it and post the published link, that's an edit link.

Should we review/merge these PRs?

**Comment by brianhenn** [comment]
[Here](https://wandb.ai/ai2cm/fourcastnet/reports/Inline-inference-speed--Vmlldzo3MzA4MzU4) it is. Based on a quick review I'm happy to merge ai2cm/full-model#714 and close ai2cm/full-model#700 and ai2cm/full-model#695, assuming we're good with the speed numbers (@oliverwm1?).

**Comment by oliverwm1** [comment]
> [Here](https://wandb.ai/ai2cm/fourcastnet/reports/Inline-inference-speed--Vmlldzo3MzA4MzU4) it is. Based on a quick review I'm happy to merge ai2cm/full-model#714 and close ai2cm/full-model#700 and ai2cm/full-model#695, assuming we're good with the speed numbers (@oliverwm1?).

Do you know what the numbers are in terms of inference steps per second?

**Comment by oliverwm1** [comment]
I guess we don't log that to wandb for training runs?

**Comment by brianhenn** [comment]
Yeah not logged for training runs (I originally caught this issue because inline inference was very slow). However I think it's correct to assume that the original "fast" inference in the report corresponds to the ~20 steps/second on Perlmutter A100 w/o data writing that we had before, so the proposed solution here being 65% as fast as that (due to the combined effects of the SeasonalAggregator and sped-up HistogramAggregator) should correspond to 13 steps/second.

**Comment by oliverwm1** [comment]
Okay got it. Well I'd say that long-term a penalty like this is not acceptable for these metrics, but for now let's just merge relevant PRs and move on I think.

For the histograms, we could potentially just compute these for every 5th timestep or something, and I suspect we would get very similar plots plots with much less cost.

Also, we could try to make computing these metrics non-blocking by doing them in a separate thread / on CPU in parallel. But that could be a lot more complicated to implement so not worth it yet IMO.

**Comment by mcgibbon** [comment]
I'm also cool with disabling the histograms in the default config (though I'll likely enable them in my own runs, for what it's worth).

> Also, we could try to make computing these metrics non-blocking by doing them in a separate thread / on CPU in parallel. But that could be a lot more complicated to implement so not worth it yet IMO.

It might not be that bad. If our performance is concerning enough that we're disabling metrics, I could try to implement this.

**Comment by brianhenn** [comment]
OK - ai2cm/full-model#695 makes `HistogramAggregator` on/off configurable for standalone inference. We don't have any configuration options for inline inference aggregators currently, so the PR also just hard-codes them to be off there, but I am OK with however we want to do that.
